django-stubs/utils/datastructures.pyi

```diff
@@ -15,61 +16,117 @@ from typing import (
+
+# Unfortunately, there's often check `if isinstance(var, (list, tuple))` in django
+# codebase. So we need sometimes to declare exactly list or tuple.
+_ListOrTuple = Union[List[_K], Tuple[_K, ...], Tuple[()]]
+
+class _PropertyDescriptor(Generic[_K, _V]):
+    """
+    This helper property descriptor allows defining asynmetric getter/setters
+    which mypy currently doesn't support with either:
+
+        class HttpResponse:
+            @property
+            def content(...): ...
+            @property.setter
+            def content(...): ...
+
+    or:
+
+        class HttpResponse:
+            def _get_content(...): ...
+            def _set_content(...): ...
+            content = property(_get_content, _set_content)
+    """
+
+    def __get__(self, instance: Any, owner: Optional[Any]) -> _V: ...
+    def __set__(self, instance: Any, value: _K) -> None: ...
+
+_IC = TypeVar("_IC", bound="_IndexableCollection")
+
+class _IndexableCollection(Protocol[_I], Collection[_I]):
+    @overload
+    def __getitem__(self, index: int) -> _I: ...
+    @overload
+    def __getitem__(self: _IC, index: slice) -> _IC: ...
 
 class OrderedSet(MutableSet[_K]):
     dict: Dict[_K, None] = ...
     def __init__(self, iterable: Optional[Iterable[_K]] = ...) -> None: ...
     def __contains__(self, item: object) -> bool: ...
     def __iter__(self) -> Iterator[_K]: ...
+    def __bool__(self) -> bool: ...
     def __len__(self) -> int: ...
-    def add(self, x: _K) -> None: ...
+    def add(self, item: _K) -> None: ...
+    def remove(self, item: _K) -> None: ...
     def discard(self, item: _K) -> None: ...
 
 class MultiValueDictKeyError(KeyError): ...
 
 _D = TypeVar("_D", bound="MultiValueDict")
 
-class MultiValueDict(MutableMapping[_K, _V]):
+class MultiValueDict(Dict[_K, _V]):
     @overload
     def __init__(self, key_to_list_mapping: Mapping[_K, Optional[List[_V]]] = ...) -> None: ...
     @overload
     def __init__(self, key_to_list_mapping: Iterable[Tuple[_K, List[_V]]] = ...) -> None: ...
-    def getlist(self, key: _K, default: Any = ...) -> List[_V]: ...
+    @overload
+    def get(self, key: _K) -> Optional[_V]: ...
+    @overload
+    def get(self, key: _K, default: _Z = ...) -> Union[_V, _Z]: ...
+    def getlist(self, key: _K, default: _Z = ...) -> Union[List[_V], _Z]: ...
     def setlist(self, key: _K, list_: List[_V]) -> None: ...
+    def setdefault(self, key: _K, default: _V = ...) -> _V: ...
     def setlistdefault(self, key: _K, default_list: Optional[List[_V]] = ...) -> List[_V]: ...
     def appendlist(self, key: _K, value: _V) -> None: ...
+    def items(self) -> Iterator[Tuple[_K, Union[_V, List[object]]]]: ...  # type: ignore
     def lists(self) -> Iterable[Tuple[_K, List[_V]]]: ...
-    def dict(self) -> Dict[_K, _V]: ...
+    def dict(self) -> Dict[_K, Union[_V, List[object]]]: ...
     def copy(self: _D) -> _D: ...
+    def __getitem__(self, key: _K) -> Union[_V, List[object]]: ...  # type: ignore
```

Comment:
One option would be to keep `MultiValueDict` as you have changed, but override `QueryDict` back to the old behavior of assuming no empty lists, for the `request.POST/GET` use case. Is that a compromise you could live with?

Although that way this issue would still affect `request.FILES`.

Reply:
Overriding `QueryDict` to old behavior could be a better choice, also I don't insist on preserving this change if you think it's troublesome and doesn't really help. 

However, I should also note that there is `QueryDict.get` interface, that works exactly like `dict.get`, returning `None or default` if key is missing or corresponds to empty list. 

Also I have one more solution, which may be a massive overhead but should provide very descriptive interface: define in stubs fake `_ImmutableQueryDict` with get methods that behave as expected (`getitem: (_K) -> _V`; it is perfectly valid assuming that this querydict was created by django and is coming from `request.GET(POST)`), modifying methods disallowed (`__setitem__: (_K, _V) -> NoReturn` etc.) and `copy: () -> QueryDict`. This way we can additionally detect almost all disallowed operations (with `--warn-unreachable`) while keeping edge-cases checked for mutable copy. It does not eliminate issue with `request.FILES`, but for this using `request.FILES.get(key)` is almost always convenient.

Reply:
Hmm, yes, the `_ImmutableQueryDict` option may work. Although it looks like there are some situations where mutation is considered legal in Django:

```python
class HttpRequest:
    def __init__(self):
        self.GET = QueryDict(mutable=True)
        self.POST = QueryDict(mutable=True)
```

Apparently this was introduced in https://github.com/django/django/pull/2778 and the reason given is "GET and POST on a vanilla HttpRequest object to be mutable QueryDicts (mutable because the Django tests, and probably many third party tests were expecting it)."

Not sure if this is relevant any more in 2022.

Reply:
Interesting, I haven't noticed that before. `HttpRequest.{GET, POST}` are really mutable, while `{W, A}SGIRequest.{GET, POST}` are not. As far as I understand (quick `grep 'HttpRequest(`), it's used directly only in tests. We don't use `{W, A}SGIRequest` in stubs because it depends on deploy method, not on code itself, so we have to make `HttpRequest.GET` immutable too in stubs. I see two options:

- add `_CommonHttpRequest(HttpRequest)` with immutable `GET/POST` and use it in stubs instead of `HttpRequest`. Will require some work and is hard to maintain.
- make `HttpRequest.{GET, POST}` immutable. It will break code that instantiates `HttpRequest` manually.

I think that the latter is a reasonable choice, but I'm not sure about it. Is there any scenario when somebody can intentionally create `HttpRequest` manually and modify its `GET/POST` data?

Reply:
I found one more solution which looks very promising. `mypy` [does not 100% support this currently](https://github.com/python/mypy/issues/1020), but the same can be achieved with a tweak (requiring `type: ignore` on `__init__` in stub, because `mypy` doesn't consider it valid, but working outside like a charm):
```
class _ImmutableQueryDict(QueryDict): ...  # declaring immutability here
class HttpRequest:
    GET: _ImmutableQueryDict
    POST: _ImmutableQueryDict
    # Magic happens here:
    def __init__(self: _MutableHttpRequest) -> None: ...  # type: ignore
    # If mypy issue #1020 is resolved, the following (better) will work:
    # def __new__(cls, *args, **kwargs) -> _MutableHttpRequest: ...
class _MutableHttpRequest(HttpRequest):
    GET: QueryDict  # type: ignore
    POST: QueryDict  # type: ignore
```

And now
```
from django.http.request import HttpRequest

request = HttpRequest()
reveal_type(request)  # N: Revealed type is "django.http.request._MutableHttpRequest"
reveal_type(request.GET)  # N: Revealed type is "django.http.request.QueryDict"
request.GET['foo'] = 'bar'

def mk_request() -> HttpRequest:
    return HttpRequest()

req = mk_request()
reveal_type(req)  # N: Revealed type is "django.http.request.HttpRequest"
reveal_type(req.GET)  # N: Revealed type is "django.http.request._ImmutableQueryDict"
req.GET['foo'] = 'bar'
```

This way any code that does for some reason `request = HttpRequest()` will get mutable version while all `request: HttpRequest` will have immutable dict.

Reply:
> If mypy issue `#1020` is resolved, the following (better) will work

Hmm, this issue https://github.com/python/mypy/issues/1020 is closed. Wasn't it solved by https://github.com/python/mypy/pull/7188 already?

Reply:
Anyway this solution is a bit hacky, but seems fine for me. Of course in the end it's up to django-stubs reviewers.

If your changes are in a good enough shape, I would gladly test out your branch (or maybe open a draft PR already).